### PR TITLE
Add nullsafe option to JsonAdapter annotation

### PR DESCRIPTION
**Pull request by chalup**

### CODE REVIEW

**Comment by chalup** [comment]
CLA signed.

**Comment by swankjesse** [comment]
I think this is a weird API. Should we just do null safety automatically with `@JsonAdapter`?

**Comment by chalup** [comment]
That would be reasonable, but strictly speaking it's a backward
incompatible change. Not sure what's the project's policy.

**Comment by swankjesse** [comment]
@inder123  ?

**Comment by swankjesse** [comment]
@JakeWharton ?

File: gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java [hunk 1]
```diff
@@ -51,18 +51,24 @@ public JsonAdapterAnnotationTypeAdapterFactory(ConstructorConstructor constructo
   static TypeAdapter<?> getTypeAdapter(ConstructorConstructor constructorConstructor, Gson gson,
       TypeToken<?> fieldType, JsonAdapter annotation) {
     Class<?> value = annotation.value();
+    TypeAdapter<?> typeAdapter;
     if (TypeAdapter.class.isAssignableFrom(value)) {
-          Class<TypeAdapter<?>> typeAdapter = (Class<TypeAdapter<?>>) value;
-      return constructorConstructor.get(TypeToken.get(typeAdapter)).construct();
-    }
-    if (TypeAdapterFactory.class.isAssignableFrom(value)) {
+          Class<TypeAdapter<?>> typeAdapterClass = (Class<TypeAdapter<?>>) value;
```
Comment (JakeWharton) [inline gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java]: nit: wrong indent

File: gson/src/main/java/com/google/gson/annotations/JsonAdapter.java [hunk 2]
```diff
@@ -94,4 +94,6 @@
   /** Either a {@link TypeAdapter} or {@link TypeAdapterFactory}. */
   Class<?> value();
 
+  /** Should the created {@link TypeAdapter} be null tolerant */
```
Comment (JakeWharton) [inline gson/src/main/java/com/google/gson/annotations/JsonAdapter.java]: nit: missing period.

**Comment by JakeWharton** [comment]
> Should we just do null safety automatically with `@JsonAdapter`?

Yeah this is probably best since it's in line with the behavior without being annotated. Or at least for fields it is. Not sure about classes in that case.

**Comment by chalup** [comment]
Nits fixed.

So we really have two questions: should nullsafe wrap be configurable and what should be the default behavior. I can't think of the non-contrived use case for not nullsafe adapter, so I'd be fine with making it non-configurable default.

Let me know what you think, and I'll update my PR.

**Comment by swankjesse** [comment]
Not configurable by default is my preference.

**Comment by swankjesse** [comment]
So please update your PR!

**Comment by chalup** [comment]
Updated.

**Comment by swankjesse** [comment]
LGTM

File: gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java [hunk 3]
```diff
@@ -51,18 +51,20 @@ public JsonAdapterAnnotationTypeAdapterFactory(ConstructorConstructor constructo
...
           .construct()
           .create(gson, fieldType);
+    } else {
+      throw new IllegalArgumentException(
+          "@JsonAdapter value must be TypeAdapter or TypeAdapterFactory reference.");
     }
 
-    throw new IllegalArgumentException(
-        "@JsonAdapter value must be TypeAdapter or TypeAdapterFactory reference.");
+    return typeAdapter.nullSafe();
```
Comment (inder123) [inline gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java]: I think we should trust the developer to do the null check.

Comment (swankjesse) [inline gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java]: The catch is that with `@JsonAdapter` there’s no easy way to do `nullSafe()`

Comment (inder123) [inline gson/src/main/java/com/google/gson/internal/bind/JsonAdapterAnnotationTypeAdapterFactory.java]: Correct. However, the developer can do the null check in code too.
Thinking again, I think it is ok for us to wrap the developer's adapter in nullSafe(). It is slightly inefficient if developer already has written the null check.

JsonSerializer/JsonDeserializer are actually similar to this as they don't require a null check.

**Pull request merged**